Bottom: 96ced6ecf93edebd6cc68e5ecccbc01969ae3430
Top:    7fdd8303aaa5995fb3983bc718b89cc65483f021
Author: Martin Kupec <martin.kupec@kupson.cz>
Date:   2010-12-23 11:41:14 +0000

Add NetByName to the select action options

v2: Check Net name appropriately
v3: Ensure we set the changed flag

Reviewed-by: Peter Clifton <pcjc2@cam.ac.uk>


---

diff --git a/src/action.c b/src/action.c
index fde1e7c..fe28c84 100644
--- a/src/action.c
+++ b/src/action.c
@@ -135,6 +135,7 @@ typedef enum
   F_Move,
   F_NameOnPCB,
   F_Netlist,
+  F_NetByName,
   F_None,
   F_Notify,
   F_Object,
@@ -371,6 +372,7 @@ static FunctionType Functions[] = {
   {"Move", F_Move},
   {"NameOnPCB", F_NameOnPCB},
   {"Netlist", F_Netlist},
+  {"NetByName", F_NetByName},
   {"None", F_None},
   {"Notify", F_Notify},
   {"Object", F_Object},
@@ -5343,8 +5345,9 @@ static const char select_syntax[] =
   "Select(All|Block|Connection)\n"
   "Select(ElementByName|ObjectByName|PadByName|PinByName)\n"
   "Select(ElementByName|ObjectByName|PadByName|PinByName, Name)\n"
-  "Select(TextByName|ViaByName)\n"
-  "Select(TextByName|ViaByName, Name)\n" "Select(Convert)";
+  "Select(TextByName|ViaByName|NetByName)\n"
+  "Select(TextByName|ViaByName|NetByName, Name)\n"
+  "Select(Convert)";
 
 static const char select_help[] = "Toggles or sets the selection";
 
@@ -5358,6 +5361,7 @@ static const char select_help[] = "Toggles or sets the selection";
 @item PinByName
 @item TextByName
 @item ViaByName
+@item NetByName
 
 These all rely on having a regular expression parser built into
 @code{pcb}.  If the name is not specified then the user is prompted
@@ -5391,7 +5395,6 @@ ActionSelect (int argc, char **argv, int x, int y)
   char *function = ARG (0);
   if (function)
     {
-
       HideCrosshair (true);
       switch (GetFunctionID (function))
 	{
@@ -5416,6 +5419,9 @@ ActionSelect (int argc, char **argv, int x, int y)
 	case F_ViaByName:
 	  type = VIA_TYPE;
 	  goto commonByName;
+	case F_NetByName:
+	  type = NET_TYPE;
+	  goto commonByName;
 
 	commonByName:
 	  {
@@ -5573,7 +5579,6 @@ static int
 ActionUnselect (int argc, char **argv, int x, int y)
 {
   char *function = ARG (0);
-
   if (function)
     {
       HideCrosshair (true);
@@ -5600,6 +5605,9 @@ ActionUnselect (int argc, char **argv, int x, int y)
 	case F_ViaByName:
 	  type = VIA_TYPE;
 	  goto commonByName;
+	case F_NetByName:
+	  type = NET_TYPE;
+	  goto commonByName;
 
 	commonByName:
 	  {
diff --git a/src/const.h b/src/const.h
index 676a6f0..e4a1d31 100644
--- a/src/const.h
+++ b/src/const.h
@@ -310,6 +310,7 @@ When set, element names are not drawn.
 #define ELEMENTARC_TYPE		0x08000
 
 #define LOCKED_TYPE 		0x10000	/* used to tell search to include locked items. */
+#define NET_TYPE		0x20000 /* used to select whole net. */
 
 #define PIN_TYPES     (VIA_TYPE | PIN_TYPE)
 #define LOCK_TYPES    (VIA_TYPE | LINE_TYPE | ARC_TYPE | POLYGON_TYPE | ELEMENT_TYPE \
diff --git a/src/select.c b/src/select.c
index 87dc661..b8d6081 100644
--- a/src/select.c
+++ b/src/select.c
@@ -44,6 +44,7 @@
 #include "undo.h"
 #include "rats.h"
 #include "misc.h"
+#include "find.h"
 
 #include <sys/types.h>
 #ifdef HAVE_REGEX_H
@@ -1011,6 +1012,33 @@ SelectObjectByName (int Type, char *Pattern, bool Flag)
       }
   }
   END_LOOP;
+  if (Type & NET_TYPE)
+    {
+      InitConnectionLookup ();
+      changed = ResetConnections (true, true) || changed;
+
+      MENU_LOOP (&PCB->NetlistLib);
+      {
+        Cardinal i;
+        LibraryEntryType *entry;
+        ConnectionType conn;
+
+        /* Name[0] and Name[1] are special purpose, not the actual name*/
+        if (menu->Name && menu->Name[0] != '\0' && menu->Name[1] != '\0' &&
+            REGEXEC (menu->Name + 2))
+          {
+            for (i = menu->EntryN, entry = menu->Entry; i; i--, entry++)
+              if (SeekPad (entry, &conn, false))
+                RatFindHook (conn.type, conn.ptr1, conn.ptr2, conn.ptr2,
+                             true, true);
+          }
+      }
+      END_LOOP;
+
+      changed = SelectConnection (Flag) || changed;
+      changed = ResetConnections (true, true) || changed;
+      FreeConnectionLookupMemory ();
+    }
 
 #if defined(HAVE_REGCOMP)
 #if !defined(sgi)
